File: frogoui/src/main/java/com/frogobox/ui/fireworks/modifiers/AlphaModifier.java
package com.frogobox.ui.fireworks.modifiers;

import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;

import com.frogobox.ui.fireworks.Particle;

public class AlphaModifier implements ParticleModifier {

    private final int mInitialValue;
    private final int mFinalValue;
    private final long mStartTime;
    private final long mEndTime;
    private final float mDuration;
    private final float mValueIncrement;
    private final Interpolator mInterpolator;

    public AlphaModifier(int initialValue, int finalValue, long startMilis, long endMilis, Interpolator interpolator) {
        mInitialValue = initialValue;
        mFinalValue = finalValue;
        mStartTime = startMilis;
        mEndTime = endMilis;
        mDuration = mEndTime - mStartTime;
        mValueIncrement = mFinalValue - mInitialValue;
        mInterpolator = interpolator;
    }

    public AlphaModifier(int initialValue, int finalValue, long startMilis, long endMilis) {
        this(initialValue, finalValue, startMilis, endMilis, new LinearInterpolator());
    }

    @Override
    public void apply(Particle particle, long miliseconds) {
        if (miliseconds < mStartTime) {
            particle.mAlpha = mInitialValue;
        } else if (miliseconds > mEndTime) {
            particle.mAlpha = mFinalValue;
        } else {
            float interpolaterdValue = mInterpolator.getInterpolation((miliseconds - mStartTime) * 1f / mDuration);
            int newAlphaValue = (int) (mInitialValue + mValueIncrement * interpolaterdValue);
            particle.mAlpha = newAlphaValue;
        }
    }

}

File: frogoui/src/main/java/com/frogobox/ui/fireworks/modifiers/AlphaModifier.kt
package com.frogobox.ui.fireworks.modifiers

import android.view.animation.Interpolator
import android.view.animation.LinearInterpolator
import com.frogobox.ui.fireworks.Particle

class AlphaModifier @JvmOverloads constructor(
    private val mInitialValue: Int,
    private val mFinalValue: Int,
    private val mStartTime: Long,
    private val mEndTime: Long,
    interpolator: Interpolator = LinearInterpolator()
) : ParticleModifier {

    private val mDuration: Float = (mEndTime - mStartTime).toFloat()
    private val mValueIncrement: Float = (mFinalValue - mInitialValue).toFloat()
    private val mInterpolator: Interpolator

    init {
        mInterpolator = interpolator
    }

    override fun apply(particle: Particle?, miliseconds: Long) {
        if (miliseconds < mStartTime) {
            particle!!.mAlpha = mInitialValue
        } else if (miliseconds > mEndTime) {
            particle!!.mAlpha = mFinalValue
        } else {
            val interpolaterdValue = mInterpolator.getInterpolation((miliseconds - mStartTime) * 1f / mDuration)
            val newAlphaValue = (mInitialValue + mValueIncrement * interpolaterdValue).toInt()
            particle!!.mAlpha = newAlphaValue
        }
    }
}